compiler/rustc_codegen_ssa/src/back/linker.rs

```diff
@@ -1557,6 +1557,48 @@ pub(crate) fn exported_symbols(tcx: TyCtxt<'_>, crate_type: CrateType) -> Vec<St
     symbols
 }
 
+pub(crate) fn linked_symbols(tcx: TyCtxt<'_>, crate_type: CrateType) -> Vec<(String, bool)> {
+    match crate_type {
+        CrateType::Executable | CrateType::Cdylib => (),
+        CrateType::Staticlib | CrateType::ProcMacro | CrateType::Rlib | CrateType::Dylib => {
+            return Vec::new();
+        }
+    }
+
+    let mut symbols = Vec::new();
+
+    let export_threshold = symbol_export::crates_export_threshold(&[crate_type]);
+    for &(symbol, info) in tcx.exported_symbols(LOCAL_CRATE).iter() {
+        if info.level.is_below_threshold(export_threshold) || info.used {
```

Comment:
Why should symbols that aren't exported be preserved?

Reply:
The linkers should be able to see these symbols, and can keep them around with `KEEP` directive, e.g. if they are placed in init_arrays section.

Reply:
But if they are placed in a `.init_arrays` section that already keeps them alive, right?

Reply:
No. Because rlibs are archives, each object file in an archive needs to satisfy at least one undefined reference to be selected for linking. This [example(https://gist.github.com/nbdd0121/0f0bae5f4e3e2b3b8695f51eee523f19) doesn't print "B" even though it's declared a ctor.

`.init_arrays` is not *that* special -- it's also kept alive by `KEEP` directive in the default linker script.

Reply:
The `.init_arrays` definition would get `#[used]`, right? So for

```
extern "C" fn my_constructor() {}

#[used]
#[link_section = ".init_array"]
static INIT_MY_CONSTRUCTOR: [extern "C" fn(); 1] = [my_constructor];
```

`my_constructor` wouldn't need to be explicitly included, but `INIT_MY_CONSTRUCTOR` would thanks to `#[used]`.

Reply:
`INIT_MY_CONSTRUCTOR` is not above the export threshold, thus the `|| info.used` part is needed to ensure `INIT_MY_CONSTRUCTOR` reaches linker.

Reply:
Yes, the `info.used` part should remain, but I would like to know why `info.level.is_below_threshold(export_threshold)` is necessary.

Reply:
If that isn't included, cdylibs would not have all symbols linked in, leading to missing exported symbols.

Reply:
Only `#[no_mangle]` symbols are exported from cdylibs. For those `info.level.is_below_threshold(export_threshold)` is false anyway.

Reply:
Why would it be false? `#[no_mangle]` symbols would be marked `SymbolExportLevel::C`, which is the lowest.

Reply:
In that case the ordering is opposite from what I expected. I thought it would return true for private symbols. Marking as resolved.